R/meltAssay.R

```diff
@@ -0,0 +1,152 @@
+#' @title meltAssay
+#'
+#' @description Converts a \code{\link[=TreeSummarizedExperiment-class]{TreeSummarizedExperiment}} object into a
+#' long data.frame which can be used for \code{\link[tidyverse]{tidyverse}}-tools.
+#'
+#' @param x a numeric matrix or a \code{\link[=TreeSummarizedExperiment-class]{TreeSummarizedExperiment}}
+#'          object containing a tree.
+#' @param abund_values Must be one of the values returned by assayNames.
+#' @param add_col_data Choice of colData columns to add. Default="none".
+#' @param ... optional arguments not used.
+#' @return A data.frame tbl_df.
+#' @examples
+#' \dontrun{
+#' data(GlobalPatterns)
+#' data(GlobalPatterns, package = "MicrobiomeExperiment")
+#' se <- GlobalPatterns
+#' se.rel <- relAbundanceCounts(se)
+#' molten_se <- meltAssay(se.rel,
+#'   add_col_data = colnames(colData(se.rel)),
+#'   abund_values = "counts"
+#' )
+#' head(molten_se)
+#' }
+#' @importFrom SummarizedExperiment assayNames rowData rowData<-
+#' @importFrom tibble rownames_to_column
+#' @importFrom tidyr pivot_longer
+#' @importFrom dplyr left_join mutate_if %>%
+#' @keywords Utilities
+#' @export
+setGeneric("meltAssay",
+  signature = "x",
+  function(x, add_col_data = "none", abund_values = "counts", ...) {
+    standardGeneric("meltAssay")
+  }
+)
+
+
+#' @rdname meltAssay
+#' @aliases meltAssay
+#'
+#' @importFrom SummarizedExperiment rowData rowData<- colData colData<-
+#'
+#' @export
+setMethod("meltAssay",
+  signature = c(x = "TreeSummarizedExperiment"),
+  function(x, add_col_data = "none", abund_values = "counts") {
+
+    # input check
+    .check_abund_values(abund_values, x)
+
+    uTaxaID <- uSamId <- Abundance <- NULL
+
+    dfSE <- .melt_assay(x, abund_values)
+    dfSE2 <- .add_taxonomic_data_to_molten_assay(x, dfSE)
+    dfSE3 <- .add_col_data_to_molten_assay(dfSE2, x, add_col_data)
+    return(dfSE3)
+  }
+)
+
+
+
+#' @description Melts assay in \code{\link[=TreeSummarizedExperiment-class]{TreeSummarizedExperiment}}
+#'        object to long data format.
+#' @param x a numeric matrix or a \code{\link[=TreeSummarizedExperiment-class]{TreeSummarizedExperiment}}
+#'          object containing a tree.
+#' @param abund_values Must be one of the values returned by assayNames.
+#'
+.melt_assay <- function(x, abund_values) {
+  uTaxaID <- NULL
+
+  # input check
+  .check_abund_values(abund_values, x)
+
+  molten_assay <- assays(x)[[abund_values]] %>%
+    data.frame() %>%
+    rownames_to_column("uTaxaID") %>%
```

Comment:
Does the `rownames` need to contain taxonomic information? I think a more general term like `uFeatureID` would be more appropriate and would be consistent with `uSamID` used below.

For what does the `u` stand for in `uTaxaID`? unique?

Maybe `SampleID` and `FeatureID` would do as well, since there are now checks for uniqueness in the code. I think the uniqueness is only derived from `rownames` and `colnames`. In addition the values are not unique in the final molten data.